### ISSUE TITLE
hash-map crashes when last argument is boolean, table or function when more than one key pair is given

### ISSUE THREAD

**Issue by rktjmp** [andreyorst/fennel-cljlib#1]
Calls such as `(hash-map :x :y :a false)` crash with a "attempt to concatenate a boolean value" error. Same for tables and functions. This only seems to occur if:

- more than one key pair is given (`(hash-map :a false)` is OK)
- the offending type is the last argument (`(hash-map :a false :x :y)` is also OK)

Behaviour is the same between Lua 5.1 and 5.4 and Fennel 1.0.0 and 1.1.0.

`nil` not included in the test as that doesn't seem to be supported anyway.

```fennel
(local {: hash-map} (require :cljlib))
(local fennel (require :fennel))

(local test-values [:string
                    false true
                    -100 0 100
                    {:assoc :table}
                    [:seq :table]
                    (fn [] true)])
(fn test [...]
  (match (pcall hash-map ...)
    (true map) (print :OK (fennel.view map))
    (false e) (print :ERROR e :arguments (fennel.view [...]))))

;; one key-value pair => OK for any type
(print "Testing one key-value pair (hash-map static-key val)")
(each [_ v (ipairs test-values)]
  (test :door v))

;; multi key-value pairs, where value is not last argument, last argument is string (to avoid bug)
(print "Testing multi key-value pair (hash-map static-key val static-key string-val)")
(each [_ v (ipairs test-values)]
  (test :door v :locked :string))

;; multi key-value pairs, where value is the last argument (bug)
(print "Testing multi key-value pair (hash-map static-key string-val static-key val)")
(each [_ v (ipairs test-values)]
  (test :locked :string :door v))
```

**output**

```
Testing one key-value pair (hash-map static-key val)
OK      {:door "string"}
OK      {:door false}
OK      {:door true}
OK      {:door -100}
OK      {:door 0}
OK      {:door 100}
OK      {:door {:assoc "table"}}
OK      {:door ["seq" "table"]}
OK      {:door #<function: 0x416fee40>}
Testing multi key-value pair (hash-map static-key val static-key string-val)
OK      {:door "string" :locked "string"}
OK      {:door false :locked "string"}
OK      {:door true :locked "string"}
OK      {:door -100 :locked "string"}
OK      {:door 0 :locked "string"}
OK      {:door 100 :locked "string"}
OK      {:door {:assoc "table"} :locked "string"}
OK      {:door ["seq" "table"] :locked "string"}
OK      {:door #<function: 0x416fee40> :locked "string"}
Testing multi key-value pair (hash-map static-key string-val static-key val)
OK      {:door "string" :locked "string"}
ERROR   ./cljlib/init.fnl:2360: attempt to concatenate a boolean value  arguments       ["locked" "string" "door" false]
ERROR   ./cljlib/init.fnl:2360: attempt to concatenate a boolean value  arguments       ["locked" "string" "door" true]
OK      {:door -100 :locked "string"}
OK      {:door 0 :locked "string"}
OK      {:door 100 :locked "string"}
ERROR   ./cljlib/init.fnl:2360: attempt to concatenate a table value    arguments       ["locked" "string" "door" {:assoc "table"}]
ERROR   ./cljlib/init.fnl:2360: attempt to concatenate a table value    arguments       ["locked" "string" "door" ["seq" "table"]]
ERROR   ./cljlib/init.fnl:2360: attempt to concatenate a function value arguments       ["locked" "string" "door" #<function: 0x416fee40>]
```

**Comment by rktjmp** [comment]
```diff
  (defn core.assoc
    "Associate key `k' with value `v' in `tbl'."
    ([tbl k v]
     (assert (not (nil? k)) "attempt to use nil as key")
     (setmetatable
      (doto tbl (tset k v))
      {:cljlib/type :table}))
    ([tbl k v & kvs]
     (assert (= (% (length kvs) 2) 0)
-            (.. "no value supplied for key " (. kvs (length kvs))))
+            (.. "no value supplied for key " (tostring (. kvs (length kvs)))))
     (assert (not (nil? k)) "attempt to use nil as key")
     (tset tbl k v)
     (var [k v] [nil nil])
     (var (i k) (next kvs))
     (while i
       (set (i v) (next kvs i))
       (tset tbl k v)
       (set (i k) (next kvs i)))
     (setmetatable tbl {:cljlib/type :table})))
```

That fixes it, debatable whether it's better to convert the `assert` to an actual `(if)` and only attempt to generate the key name if required? Might still be worth retaining the `(tostring)` call as keys might be non-strings.

```diff
  (defn core.assoc
    "Associate key `k' with value `v' in `tbl'."
    ([tbl k v]
     (assert (not (nil? k)) "attempt to use nil as key")
     (setmetatable
      (doto tbl (tset k v))
      {:cljlib/type :table}))
    ([tbl k v & kvs]
+    (when (not (= (% (length kvs) 2) 0))
+      (error (.. "no value supplied for key " (tostring (. kvs (length kvs))))))
-    (assert (= (% (length kvs) 2) 0)
-            (.. "no value supplied for key " (. kvs (length kvs))))
     (assert (not (nil? k)) "attempt to use nil as key")
     (tset tbl k v)
     (var [k v] [nil nil])
     (var (i k) (next kvs))
     (while i
       (set (i v) (next kvs i))
       (tset tbl k v)
       (set (i k) (next kvs i)))
     (setmetatable tbl {:cljlib/type :table})))
```

**Comment by andreyorst** [comment]
should be fixed now with f48aa44ba3bdd669c539afb4bb2881e332b77dbe. Thanks for catching this!
